_posts/2020-01-27-paging-and-sorting-in-distributed-systems-oh-my.md

```diff
@@ -0,0 +1,137 @@
+
+>"we could say that *Marketing/Product Catalog* is the one that makes decisions about the existence of a Product. Which means that from the business perspective, *Marketing* owns the concept of a Product."
+
+In that article, we briefly identified what the process of composing a list is:
+
+* *Marketing* loads the products list that, in the simplest case, is only a list of IDs.
+* *Marketing* publishes an event to notify interested services that they can proceed with the composition process.
+
+>Note: More details about the entire composition process are available in [The ViewModels Lists Composition Dance](/view-model-composition/2019/03/21/the-viewmodels-lists-composition-dance.html) article.
+
+## Paging
+If *Marketing* is the logical owner of products and is also the one responsible for loading the products list, *Marketing* can easily be responsible for paging too.
+
+When a request, such as `/products?pageIndex=0&pageSize=10`, comes in *Marketing* uses the paging information to restrict the number of IDs loaded, and if required, to calculate the total amount of pages.
+
+All the other services receive, in the event raised by *Marketing*, an already paged/trimmed list of products, making so that they do not need to be aware of any paging concern at all.
+
+## Sorting
+Sorting, unfortunately, it's not so easy as paging. For simple sorting scenarios, like the one in which we need to sort by a single property, the approach might be straightforward. Might be.
+
+When a request like `/products?orderBy=price` comes in what could happen is that:
+
+* *Marketing* behaves the usual way, loading product IDs and publishing the corresponding event.
+* All other services, but *Sales*, react as usual loading data by product ID and composing the products list.
+* *Sales* loads products' prices, and once its part of the composition is complete, it sorts the list by price. *Sales* is the one that can handle the `orderBy` query string parameter when the value is `price`.
+
+Unfortunately, such a simple approach immediately fails if services are invoked in parallel, like in our case. A service cannot directly change the order of something that is simultaneously touched by other services.
+
+If the request coming in is like `/products?orderBy=price,description` then sorting is not anymore so easy, even if we were allowed to sort the list directly. The problem, in this second scenario, is that *Sales* can sort by price, *Marketing* can sort by description, but what the request is asking for is different:
+
+>I want products sorted by price and then by description.
+
+Which, in essence, creates a kind of circular dependency between *Sales* and *Marketing*. *Marketing* must execute after *Sales*. Worse, it needs to know who already sorted data and how because of the "and then by" sort dependency. There is no way we can achieve this as *Marketing*, being the logical owner, is already executed first.
```

Comment:
```suggestion
In essence, this creates a kind of circular dependency between *Sales* and *Marketing*. *Marketing* must execute after *Sales*. Worse, it needs to know who already sorted data and how because of the "and then by" sort dependency. There is no way we can achieve this as the *Marketing* endpoint, being the logical owner, has already executed.
```